bridge/github/import.go

```diff
@@ -15,94 +15,155 @@ import (
 	"github.com/MichaelMure/git-bug/util/text"
 )
 
+const EMPTY_TITLE_PLACEHOLDER = "<empty string>"
+
 // githubImporter implement the Importer interface
 type githubImporter struct {
 	conf core.Configuration
 
-	// default client
-	client *githubv4.Client
-
-	// iterator
-	iterator *iterator
+	// mediator to access the Github API
+	mediator *importMediator
 
 	// send only channel
 	out chan<- core.ImportResult
 }
 
-func (gi *githubImporter) Init(_ context.Context, repo *cache.RepoCache, conf core.Configuration) error {
+func (gi *githubImporter) Init(_ context.Context, _ *cache.RepoCache, conf core.Configuration) error {
 	gi.conf = conf
+	return nil
+}
 
+// ImportAll iterate over all the configured repository issues and ensure the creation of the
+// missing issues / timeline items / edits / label events ...
+func (gi *githubImporter) ImportAll(ctx context.Context, repo *cache.RepoCache, since time.Time) (<-chan core.ImportResult, error) {
 	creds, err := auth.List(repo,
 		auth.WithTarget(target),
 		auth.WithKind(auth.KindToken),
-		auth.WithMeta(auth.MetaKeyLogin, conf[confKeyDefaultLogin]),
+		auth.WithMeta(auth.MetaKeyLogin, gi.conf[confKeyDefaultLogin]),
 	)
 	if err != nil {
-		return err
+		return nil, err
 	}
-
-	if len(creds) == 0 {
-		return ErrMissingIdentityToken
+	if len(creds) <= 0 {
+		return nil, ErrMissingIdentityToken
 	}
-
-	gi.client = buildClient(creds[0].(*auth.Token))
-
-	return nil
-}
-
-// ImportAll iterate over all the configured repository issues and ensure the creation of the
-// missing issues / timeline items / edits / label events ...
-func (gi *githubImporter) ImportAll(ctx context.Context, repo *cache.RepoCache, since time.Time) (<-chan core.ImportResult, error) {
-	gi.iterator = NewIterator(ctx, gi.client, 10, gi.conf[confKeyOwner], gi.conf[confKeyProject], since)
+	client := buildClient(creds[0].(*auth.Token))
+	gi.mediator = NewImportMediator(ctx, client, gi.conf[confKeyOwner], gi.conf[confKeyProject], since)
 	out := make(chan core.ImportResult)
 	gi.out = out
 
 	go func() {
 		defer close(gi.out)
-
-		// Loop over all matching issues
-		for gi.iterator.NextIssue() {
-			issue := gi.iterator.IssueValue()
-			// create issue
-			b, err := gi.ensureIssue(repo, issue)
-			if err != nil {
-				err := fmt.Errorf("issue creation: %v", err)
-				out <- core.NewImportError(err, "")
-				return
+		var currBug *cache.BugCache
+		var currEvent ImportEvent
+		var nextEvent ImportEvent
+		var err error
+		for {
+			// We need the current event and one look ahead event.
```

Comment:
nitpicking: it'd be great if you could explain *why* this structure with look ahead is necesary

Reply:
Before I add the comment to the code I would like to ask you for additional input: The reason for using a look ahead element is that in case a github issue is followed by a github issue edit, then in the call to `NewBugRaw` we use both, the github issue and the first github issue edit, to create the bug instance. This behaviour originates from the previous versions of the github bridge and I adopted it. It seems that this is due to the way github serves its issues on the API. (The same is true for comments.) My question is the following: Is this behaviour necessary? Would it be okay to simply create a (possibly empty) bug instance and then apply the first github issue edit in order to create the issue/bug content. This would cause the edit-history of a github issue and a corresponding git-bug bug instance to not be equal (one might display without "edit" and the other might display with "edit"). I assume that would not be acceptable - right?

Reply:
Before I add the comment to the code I would like to ask you for additional input: The reason for using a look ahead element is that in case a github issue is followed by a github issue edit, then in the call to `NewBugRaw` we use both, the github issue and the first github issue edit, to create the bug instance. This behaviour originates from the previous versions of the github bridge and I adopted it. It seems that this is due to the way github serves its issues on the API. (The same is true for comments.) My question is the following: Is this behaviour necessary? Would it be okay to simply create a (possibly empty) bug instance and then apply the first github issue edit in order to create the issue/bug content. This would cause the edit-history of a github issue and a corresponding git-bug bug instance to not be equal (one might display without "edit" and the other might display with "edit"). I assume that would not be acceptable - right?

Reply:
Isn't that because on github it's sort of as if the initial creation of an issue is (creation+comment), while in git-bug it's the same operation? I don't remember exactly.

>  I assume that would not be acceptable - right?

At least not desirable. But your code does that properly at the moment, no?

Just a thought, but there might be a way to simplify. Instead of maintaining two two values `currEvent` and `nextEvent`:
- simply iterate on the event stream
- when you know you need an extra event, consume an extra one

Reply:
>Isn't that because on github it's sort of as if the initial creation of an issue is (creation+comment), while in git-bug it's the same operation? I don't remember exactly.

In github the issue holds the most recent content. Because we want to imitate the history, we sometimes create the bug instance from the plain github issue (if there are no edits in github) and sometimes we create the bug from the github issue together with its first edit (if there are edits in github).
But that's also the reason why we do not know whether or not we need the next event before we looked at it.

>> I assume that would not be acceptable - right?
>
> At least not desirable. But your code does that properly at the moment, no?

Yes, it is handled properly.

If you want to get rid of the look ahead, then I think it would be best to make `importMediator` provide more context and add a field (say `hasEdits bool`) to `IssueEvent`. Then we just know whether or not there are edits and we do not have to look ahead. (And we would have to add a similar data field to `TimelineEvent`, because it holds comments which might be followed by comment edits. It's the same situation.)

For now I only added a comment with ede5e21 explaining why look ahead is necessary. Let me know what you think.